adapters/repos/db/vector/cache/paginated_cache.go

```diff
@@ -0,0 +1,505 @@
+		err := update(page, idx)
+		s.shardedLocks.Unlock(id)
+		return err
+	}
+	s.shardedLocks.Unlock(id)
+
+	s.pageLock.Lock()
+	defer s.pageLock.Unlock()
+
+	// try again in case the page was created in the meantime.
+	// this is to avoid the waiting routines to all acquire a LockAll
+	// lock when multiple goroutines try to create the same page.
+	s.shardedLocks.Lock(id)
+	page, idx = s.getPageForID(id)
+	if page != nil {
+		err := update(page, idx)
+		s.shardedLocks.Unlock(id)
+		return err
+	}
+	s.shardedLocks.Unlock(id)
+
+	// try again, this time with all locks held
+	s.shardedLocks.LockAll()
+	page, idx = s.getPageForID(id)
+	if page != nil {
+		err := update(page, idx)
+		s.shardedLocks.UnlockAll()
+		return err
+	}
+
+	// page doesn't exist yet, create it
+	pageIdx := id / PageSize
+	page = make([][]T, PageSize)
+	s.cache[pageIdx] = page
+	idx = int(id % PageSize)
+	err := update(page, idx)
+
+	// opportunistically create the next page
+	pageIdx++
+	if len(s.cache) > int(pageIdx) && s.cache[pageIdx] == nil {
+		page = make([][]T, PageSize)
+		s.cache[pageIdx] = page
+	}
+
+	s.shardedLocks.UnlockAll()
+	return err
+}
+
+func (s *paginatedCache[T]) Get(ctx context.Context, id uint64) ([]T, error) {
+	var vec []T
+
+	s.shardedLocks.RLock(id)
+	page, idx := s.getPageForID(id)
+	if page != nil {
+		vec = page[idx]
+	}
+	s.shardedLocks.RUnlock(id)
+
+	if vec != nil {
+		return vec, nil
+	}
+
+	return s.handleCacheMiss(ctx, id)
+}
+
+func (s *paginatedCache[T]) Delete(ctx context.Context, id uint64) {
+	s.shardedLocks.Lock(id)
+	defer s.shardedLocks.Unlock(id)
+
+	if int(id) >= len(s.cache)*PageSize {
+		return
+	}
+
+	page, idx := s.getPageForID(id)
+	if page == nil || page[idx] == nil {
+		return
+	}
+
+	page[idx] = nil
+	atomic.AddInt64(&s.count, -1)
+}
+
+func (s *paginatedCache[T]) handleCacheMiss(ctx context.Context, id uint64) ([]T, error) {
+	vec, err := s.vectorForID(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	atomic.AddInt64(&s.count, 1)
+
+	err = s.upsert(id, func(page [][]T, idx int) error {
+		page[idx] = vec
+		return nil
+	})
+
+	return vec, err
+}
+
+func (s *paginatedCache[T]) MultiGet(ctx context.Context, ids []uint64) ([][]T, []error) {
+	out := make([][]T, len(ids))
+	errs := make([]error, len(ids))
+
+	for i, id := range ids {
+		out[i], errs[i] = s.Get(ctx, id)
+	}
+
+	return out, errs
+}
+
+var prefetchFunc func(in uintptr) = func(in uintptr) {
+	// do nothing on default arch
+	// this function will be overridden for amd64
+}
+
+func (s *paginatedCache[T]) Prefetch(id uint64) {
+	s.shardedLocks.RLock(id)
+	page, idx := s.getPageForID(id)
+	if page != nil {
+		prefetchFunc(uintptr(unsafe.Pointer(&page[idx])))
+	}
+	s.shardedLocks.RUnlock(id)
+}
+
+func (s *paginatedCache[T]) Preload(id uint64, vec []T) {
+	s.shardedLocks.Lock(id)
+
+	page, idx := s.getPageForID(id)
+	if page == nil {
+		s.shardedLocks.Unlock(id)
+		return
+	}
+
+	page[idx] = vec
+	s.shardedLocks.Unlock(id)
+
+	atomic.AddInt64(&s.count, 1)
```

Comment:
Shouldn't be updated only if vector was missing in the cache?
Multiple calls with for the same `id` will increment `count` multiple times.